src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,267 @@
+                    // In this case, choose the subplan with the lowest candidate from subPlansWithStatistics as candidate plan
+                    tmpCandidateSubPlans = Arrays.asList( PhysicalPlanWithStatisticsUtils.chooseFirstSubPlan(subPlansWithStatistics) );
+                }
+
+                // Keep candidate SubPlans that use the same vocabulary mapping with the previous selected SubPlan.
+                // If 'candidateSubPlans' is empty, use all subplans in 'tmpCandidateSubPlans' as candidate subplans.
+                List<PhysicalPlanWithStatistics>  candidateSubPlans = PhysicalPlanWithStatisticsUtils.getSubPlansWithSameVoc( nextSelectedSubPlan.getVocabularyMappings(), tmpCandidateSubPlans);
+                if( candidateSubPlans.isEmpty() ){
+                    candidateSubPlans = tmpCandidateSubPlans;
+                }
+
+                // The candidateSubPlans are added to the ordered list 'orderedCandidateSubPlans'
+                orderedCandidateSubPlans.addAll(candidateSubPlans);
+
+                // select the first element (with the lowest cardinality) from orderedCandidateSubPlans as the nextSelectedSubPlan
+                // Remove the selected subPlan from subPlansWithStatistics and clear orderedCandidateSubPlans for the next loop
+                nextSelectedSubPlan = orderedCandidateSubPlans.poll();
+                subPlansWithStatistics.remove(nextSelectedSubPlan);
+                orderedCandidateSubPlans.clear();
+
+                // Construct a physical plan: two physical algorithms can be selected by calling following functions.
+//                currentJoinPlan = constructPlanWithSHJ( currentJoinPlan, nextSelectedSubPlan.plan );
+                currentJoinPlan = constructPlanWithBJ( currentJoinPlan, nextSelectedSubPlan.plan );
+            }
+
+            return currentJoinPlan;
+        }
+
+        /**
+         * @param resps A list of cardinality, which is ordered in the order in which the request operators can be found by a depth-first traversal of the subplans.
+         * @return A list of PhysicalPlanWithStatistics, with each object corresponding to a subplan in the 'subplans' list.
+         * Each PhysicalPlanWithStatistics object contains the physical plan for the subplan, along with its cardinality,
+         * a list of relevant federations members, and estimated number of access to federation members.
+         *
+         * During a depth-first traversal of the subplans, increase the index by 1 when visiting a request operator:
+         * If the root operator of the subplan is a Request operator, get the corresponding cardinality from resps with the same index;
+         * If the subPlan with a Union of requests, calculate the total cardinality by summing up the cardinality of individual requests.
+         * (See page 1052 in the paper[1])
+         */
+        protected List<PhysicalPlanWithStatistics> associateCardWithSubPlans(
```

Comment:
Is this the same function as in `CardinalityBasedGreedyJoinPlanOptimizerImpl`? If so, can you reuse it instead of copying the code?